### ISSUE TITLE
3.3.1: ${prefix} missing in /usr/lib/pkg-config

### ISSUE THREAD

**Issue by Ra2-IFV** [openssl/openssl#24727]
I was trying to update openssl for openwrt, from `3.0.14` to `3.3.1`. I found `${prefix}` is missing in `includedir` and `libdir`, in `/usr/lib/pkg-config/openssl.pc`. This causes other packages to use the wrong openssl library(not sure about this).  
How to solve this? or am I wrong?

### 3.0.14

Configure
```
(cd /home/debian/openwrt/build_dir/target-aarch64_generic_musl/openssl-3.0.14; ./Configure linux-aarch64-openwrt --prefi
x=/usr --libdir=lib --openssldir=/etc/ssl --cross-compile-prefix="aarch64-openwrt-linux-musl-" -march=armv8-a -mtune=cor
tex-a72.cortex-a53 -fno-caller-saves -fno-plt -pipe -fhonour-copts -fmacro-prefix-map=/home/debian/openwrt/build_dir/tar
get-aarch64_generic_musl/openssl-3.0.14=openssl-3.0.14 -ffunction-sections -fdata-sections -Wformat -Werror=format-secur
ity -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -O3 -DPIC -fPIC -I/home/debian/openwrt/staging_dir/too
lchain-aarch64_generic_gcc-14.1.0_musl/usr/include -I/home/debian/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1
.0_musl/include -I/home/debian/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1.0_musl/include/fortify -L/home/deb
ian/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1.0_musl/usr/lib -L/home/debian/openwrt/staging_dir/toolchain-a
arch64_generic_gcc-14.1.0_musl/lib -fuse-ld=mold -Wl,--gc-sections -znow -zrelro shared no-tests no-blake2 -DOPENSSL_PRE
FER_CHACHA_OVER_GCM no-async no-ec2m no-aria no-sm2 no-sm3 no-sm4 no-camellia no-idea no-seed no-mdc2 no-whirlpool no-rf
c3779 no-engine no-dtls no-comp no-nextprotoneg && { [ -f /home/debian/openwrt/build_dir/target-aarch64_generic_musl/ope
nssl-3.0.14/.configured_4c5d84187db499b3d30610b767286fb0_3270042c5cbe4bfdcf0814af95a8a8fb ] || make clean; } )
```
openssl.pc
```
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.0.14
Requires: libssl libcrypto
```

[3.0.14_configdata.pm.txt](https://github.com/user-attachments/files/15964259/3.0.14_configdata.pm.txt)  
No `includedata.pm`.

### 3.3.1

Configure
```
(cd /home/debian/openwrt/build_dir/target-aarch64_generic_musl/openssl-3.3.1; ./Configure linux-aarch64-openwrt --prefix
=/usr --libdir=lib --openssldir=/etc/ssl --cross-compile-prefix="aarch64-openwrt-linux-musl-" -march=armv8-a -mtune=cort
ex-a72.cortex-a53 -fno-caller-saves -fno-plt -pipe -fhonour-copts -fmacro-prefix-map=/home/debian/openwrt/build_dir/targ
et-aarch64_generic_musl/openssl-3.3.1=openssl-3.3.1 -ffunction-sections -fdata-sections -Wformat -Werror=format-security
 -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -O3 -DPIC -fPIC -I/home/debian/openwrt/staging_dir/toolch
ain-aarch64_generic_gcc-14.1.0_musl/usr/include -I/home/debian/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1.0_
musl/include -I/home/debian/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1.0_musl/include/fortify -L/home/debian
/openwrt/staging_dir/toolchain-aarch64_generic_gcc-14.1.0_musl/usr/lib -L/home/debian/openwrt/staging_dir/toolchain-aarc
h64_generic_gcc-14.1.0_musl/lib -fuse-ld=mold -Wl,--gc-sections -znow -zrelro shared no-tests no-blake2 -DOPENSSL_PREFER
_CHACHA_OVER_GCM no-async no-ec2m no-aria no-sm2 no-sm3 no-sm4 no-camellia no-idea no-seed no-mdc2 no-whirlpool no-rfc37
79 no-engine no-dtls no-comp no-nextprotoneg && { [ -f /home/debian/openwrt/build_dir/target-aarch64_generic_musl/openss
l-3.3.1/.configured_4c5d84187db499b3d30610b767286fb0_3270042c5cbe4bfdcf0814af95a8a8fb ] || make clean; } )
```
openssl.pc
```
libdir=/usr/lib
includedir=/usr/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.1
Requires: libssl libcrypto
```

[3.3.1_configdata.pm.txt](https://github.com/user-attachments/files/15964307/3.3.1_configdata.pm.txt)  
installdata.pm
```
package OpenSSL::safe::installdata;

use strict;
use warnings;
use Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw($PREFIX
                  $BINDIR $BINDIR_REL
                  $LIBDIR $LIBDIR_REL
                  $INCLUDEDIR $INCLUDEDIR_REL
                  $APPLINKDIR $APPLINKDIR_REL
                  $ENGINESDIR $ENGINESDIR_REL
                  $MODULESDIR $MODULESDIR_REL
                  $PKGCONFIGDIR $PKGCONFIGDIR_REL
                  $CMAKECONFIGDIR $CMAKECONFIGDIR_REL
                  $VERSION @LDLIBS);

our $PREFIX             = '/usr';
our $BINDIR             = '/usr/bin';
our $BINDIR_REL         = 'bin';
our $LIBDIR             = '/usr/lib';
our $LIBDIR_REL         = 'lib';
our $INCLUDEDIR         = '/usr/include';
our $INCLUDEDIR_REL     = 'include';
our $APPLINKDIR         = '/usr/include/openssl';
our $APPLINKDIR_REL     = 'include/openssl';
our $ENGINESDIR         = '/usr/lib/engines-3';
our $ENGINESDIR_REL     = 'lib/engines-3';
our $MODULESDIR         = '/usr/lib/ossl-modules';
our $MODULESDIR_REL     = 'lib/ossl-modules';
our $PKGCONFIGDIR       = '/usr/lib/pkgconfig';
our $PKGCONFIGDIR_REL   = 'lib/pkgconfig';
our $CMAKECONFIGDIR     = '/usr/lib/cmake/OpenSSL';
our $CMAKECONFIGDIR_REL = 'lib/cmake/OpenSSL';
our $VERSION            = '3.3.1';
our @LDLIBS             =
    # Unix and Windows use space separation, VMS uses comma separation
    split(/ +| *, */, '-ldl -pthread ');

1;
```

**Comment by Ra2-IFV** [comment]
```
prefix=/usr
exec_prefix=${prefix}
libdir=
includedir=${prefix}/

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.1
Requires: libssl libcrypto
```
Have to refresh https://github.com/openssl/openssl/pull/24687/commits/26d86382ea9ed144cae407437a2ac266f9146eee to patch files successfully. Seems it's still broken...

**Comment by levitte** [comment]
I did a quick test to see if I can reproduce that result.

In a checkout of `openssl-3.3` (top commit is 32aa3d13fae1ac50f85733626108688a76e98f84), I cherry-picked #24687:

``` console
$ git cherry-pick -x 26d86382ea9ed144cae407437a2ac266f9146eee^..fix-24298
[openssl-3.3 e10ac5b527] Give util/mkinstallvars.pl more fine grained control over var dependencies
 Date: Thu Jun 20 14:30:16 2024 +0200
 2 files changed, 85 insertions(+), 50 deletions(-)
[openssl-3.3 8ead0be2c9] Adapt all the exporter files to the new vars from util/mkinstallvars.pl
 Date: Thu Jun 20 14:33:15 2024 +0200
 4 files changed, 24 insertions(+), 11 deletions(-)
[openssl-3.3 258d4c25ea] fixup! Adapt all the exporter files to the new vars from util/mkinstallvars.pl
 Date: Thu Jun 20 18:49:35 2024 +0200
 1 file changed, 1 insertion(+), 1 deletion(-)
```

The configured and built (well actually, only generated), using the prefix, libdir and openssldir values you've displayed:

``` console
$ mkdir _build
$ cd _build/
$ ../Configure --prefix=/usr --libdir=lib --openssldir=/etc/ssl
Configuring OpenSSL version 3.3.2-dev for target linux-x86_64
Using os-specific seed configuration
Created configdata.pm
Running configdata.pm
Created Makefile.in
Created Makefile
Created include/openssl/configuration.h

**********************************************************************
***                                                                ***
***   OpenSSL has been successfully configured                     ***
***                                                                ***
***   If you encounter a problem while building, please open an    ***
***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***
***   and include the output from the following command:           ***
***                                                                ***
***       perl configdata.pm --dump                                ***
***                                                                ***
***   (If you are new to OpenSSL, you might want to consult the    ***
***   'Troubleshooting' section in the INSTALL.md file first)      ***
***                                                                ***
**********************************************************************
$ make build_generated
/usr/bin/perl ../util/mkinstallvars.pl PREFIX=. BINDIR=apps LIBDIR= INCLUDEDIR=include APPLINKDIR=ms ENGINESDIR=engines MODULESDIR=providers "VERSION=3.3.2-dev" "LDLIBS=-ldl -pthread " > builddata.pm
DEBUG: all keys: APPLINKDIR, BINDIR, CMAKECONFIGDIR, ENGINESDIR, INCLUDEDIR, LDLIBS, LIBDIR, MODULESDIR, PKGCONFIGDIR, PREFIX, VERSION, libdir
No value given for CMAKECONFIGDIR
No value given for PKGCONFIGDIR
No value given for libdir
DEBUG: PREFIX = . => PREFIX = /home/levitte/gitwrk/OpenSSL/official/3.3/_build
DEBUG: libdir = . => libdir = /home/levitte/gitwrk/OpenSSL/official/3.3/_build
DEBUG: BINDIR = apps => BINDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build/apps ,  BINDIR_REL_PREFIX = apps
DEBUG: LIBDIR = . => LIBDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build ,  LIBDIR_REL_PREFIX = .
DEBUG: INCLUDEDIR = include => INCLUDEDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build/include ,  INCLUDEDIR_REL_PREFIX = include
DEBUG: APPLINKDIR = ms => APPLINKDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build/ms ,  APPLINKDIR_REL_PREFIX = ms
DEBUG: ENGINESDIR = engines => ENGINESDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build/engines ,  ENGINESDIR_REL_LIBDIR = engines
DEBUG: MODULESDIR = providers => MODULESDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build/providers ,  MODULESDIR_REL_LIBDIR = providers
DEBUG: PKGCONFIGDIR = . => PKGCONFIGDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build ,  PKGCONFIGDIR_REL_LIBDIR = .
DEBUG: CMAKECONFIGDIR = . => CMAKECONFIGDIR = /home/levitte/gitwrk/OpenSSL/official/3.3/_build ,  CMAKECONFIGDIR_REL_LIBDIR = .
/usr/bin/perl "-I." "-Mconfigdata" "-Mbuilddata" "../util/dofile.pl" "-oMakefile" ../exporters/cmake/OpenSSLConfig.cmake.in > OpenSSLConfig.cmake
/usr/bin/perl "-I." "-Mconfigdata" "-Mconfigdata" "-Mbuilddata" "../util/dofile.pl" "-oMakefile" ../exporters/cmake/OpenSSLConfigVersion.cmake.in > OpenSSLConfigVersion.cmake
/usr/bin/perl "-I." "-I../util/perl" "-Mconfigdata" "-MOpenSSL::paramnames" "../util/dofile.pl" "-oMakefile" ../crypto/params_idx.c.in > crypto/params_idx.c
/usr/bin/perl ../util/mkinstallvars.pl "PREFIX=/usr" BINDIR=bin "LIBDIR=lib" "libdir=/usr/lib" INCLUDEDIR=include APPLINKDIR=include/openssl "ENGINESDIR=/usr/lib/engines-3" "MODULESDIR=/usr/lib/ossl-modules" "PKGCONFIGDIR=/usr/lib/pkgconfig" "CMAKECONFIGDIR=/usr/lib/cmake/OpenSSL" "LDLIBS=-ldl -pthread " "VERSION=3.3.2-dev" > installdata.pm
DEBUG: all keys: APPLINKDIR, BINDIR, CMAKECONFIGDIR, ENGINESDIR, INCLUDEDIR, LDLIBS, LIBDIR, MODULESDIR, PKGCONFIGDIR, PREFIX, VERSION, libdir
DEBUG: PREFIX = /usr => PREFIX = /usr
DEBUG: libdir = /usr/lib => libdir = /usr/lib
DEBUG: BINDIR = bin => BINDIR = /usr/bin ,  BINDIR_REL_PREFIX = bin
DEBUG: LIBDIR = lib => LIBDIR = /usr/lib ,  LIBDIR_REL_PREFIX = lib
DEBUG: INCLUDEDIR = include => INCLUDEDIR = /usr/include ,  INCLUDEDIR_REL_PREFIX = include
DEBUG: APPLINKDIR = include/openssl => APPLINKDIR = /usr/include/openssl ,  APPLINKDIR_REL_PREFIX = include/openssl
DEBUG: ENGINESDIR = /usr/lib/engines-3 => ENGINESDIR_REL_LIBDIR = engines-3
DEBUG: MODULESDIR = /usr/lib/ossl-modules => MODULESDIR_REL_LIBDIR = ossl-modules
DEBUG: PKGCONFIGDIR = /usr/lib/pkgconfig => PKGCONFIGDIR_REL_LIBDIR = pkgconfig
DEBUG: CMAKECONFIGDIR = /usr/lib/cmake/OpenSSL => CMAKECONFIGDIR_REL_LIBDIR = cmake/OpenSSL
/usr/bin/perl "-I." "-Mconfigdata" "-Minstalldata" "../util/dofile.pl" "-oMakefile" ../exporters/cmake/OpenSSLConfig.cmake.in > exporters/OpenSSLConfig.cmake
/usr/bin/perl "-I." "-Mconfigdata" "-Mconfigdata" "-Minstalldata" "../util/dofile.pl" "-oMakefile" ../exporters/cmake/OpenSSLConfigVersion.cmake.in > exporters/OpenSSLConfigVersion.cmake
/usr/bin/perl "-I." "-Mconfigdata" "-Minstalldata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/libcrypto.pc.in > exporters/libcrypto.pc
/usr/bin/perl "-I." "-Mconfigdata" "-Minstalldata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/libssl.pc.in > exporters/libssl.pc
/usr/bin/perl "-I." "-Mconfigdata" "-Mconfigdata" "-Mconfigdata" "-Minstalldata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/openssl.pc.in > exporters/openssl.pc
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/crypto/bn_conf.h.in > include/crypto/bn_conf.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/crypto/dso_conf.h.in > include/crypto/dso_conf.h
/usr/bin/perl "-I." "-I../util/perl" "-Mconfigdata" "-MOpenSSL::paramnames" "../util/dofile.pl" "-oMakefile" ../include/internal/param_names.h.in > include/internal/param_names.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/asn1.h.in > include/openssl/asn1.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/asn1t.h.in > include/openssl/asn1t.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/bio.h.in > include/openssl/bio.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/cmp.h.in > include/openssl/cmp.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/cms.h.in > include/openssl/cms.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/conf.h.in > include/openssl/conf.h
/usr/bin/perl "-I." "-I../util/perl" "-Mconfigdata" "-MOpenSSL::paramnames" "../util/dofile.pl" "-oMakefile" ../include/openssl/core_names.h.in > include/openssl/core_names.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/crmf.h.in > include/openssl/crmf.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/crypto.h.in > include/openssl/crypto.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/ct.h.in > include/openssl/ct.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/err.h.in > include/openssl/err.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/ess.h.in > include/openssl/ess.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/fipskey.h.in > include/openssl/fipskey.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/lhash.h.in > include/openssl/lhash.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/ocsp.h.in > include/openssl/ocsp.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/opensslv.h.in > include/openssl/opensslv.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/pkcs12.h.in > include/openssl/pkcs12.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/pkcs7.h.in > include/openssl/pkcs7.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/safestack.h.in > include/openssl/safestack.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/srp.h.in > include/openssl/srp.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/ssl.h.in > include/openssl/ssl.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/ui.h.in > include/openssl/ui.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/x509.h.in > include/openssl/x509.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/x509_vfy.h.in > include/openssl/x509_vfy.h
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../include/openssl/x509v3.h.in > include/openssl/x509v3.h
/usr/bin/perl "-I." "-Mconfigdata" "-Mbuilddata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/libcrypto.pc.in > libcrypto.pc
/usr/bin/perl "-I." "-Mconfigdata" "-Mbuilddata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/libssl.pc.in > libssl.pc
/usr/bin/perl "-I." "-Mconfigdata" "-Mbuilddata" "-Mbuilddata" "-Mbuilddata" "../util/dofile.pl" "-oMakefile" ../exporters/pkg-config/openssl.pc.in > openssl.pc
/usr/bin/perl "-I." "-Mconfigdata" "../util/dofile.pl" "-oMakefile" ../test/provider_internal_test.cnf.in > test/provider_internal_test.cnf
```

These are the resulting files to be installed:

``` console
$ cat installdata.pm 
package OpenSSL::safe::installdata;

use strict;
use warnings;
use Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(
    $PREFIX
    $libdir
    $BINDIR $BINDIR_REL_PREFIX
    $LIBDIR $LIBDIR_REL_PREFIX
    $INCLUDEDIR $INCLUDEDIR_REL_PREFIX
    $APPLINKDIR $APPLINKDIR_REL_PREFIX
    $ENGINESDIR $ENGINESDIR_REL_LIBDIR
    $MODULESDIR $MODULESDIR_REL_LIBDIR
    $PKGCONFIGDIR $PKGCONFIGDIR_REL_LIBDIR
    $CMAKECONFIGDIR $CMAKECONFIGDIR_REL_LIBDIR
    $VERSION @LDLIBS
);

our $PREFIX                     = '/usr';
our $libdir                     = '/usr/lib';
our $BINDIR                     = '/usr/bin';
our $BINDIR_REL_PREFIX          = 'bin';
our $LIBDIR                     = '/usr/lib';
our $LIBDIR_REL_PREFIX          = 'lib';
our $INCLUDEDIR                 = '/usr/include';
our $INCLUDEDIR_REL_PREFIX      = 'include';
our $APPLINKDIR                 = '/usr/include/openssl';
our $APPLINKDIR_REL_PREFIX      = 'include/openssl';
our $ENGINESDIR                 = '/usr/lib/engines-3';
our $ENGINESDIR_REL_LIBDIR      = 'engines-3';
our $MODULESDIR                 = '/usr/lib/ossl-modules';
our $MODULESDIR_REL_LIBDIR      = 'ossl-modules';
our $PKGCONFIGDIR               = '/usr/lib/pkgconfig';
our $PKGCONFIGDIR_REL_LIBDIR    = 'pkgconfig';
our $CMAKECONFIGDIR             = '/usr/lib/cmake/OpenSSL';
our $CMAKECONFIGDIR_REL_LIBDIR  = 'cmake/OpenSSL';
our $VERSION                    = '3.3.2-dev';
our @LDLIBS                     =
    # Unix and Windows use space separation, VMS uses comma separation
    split(/ +| *, */, '-ldl -pthread ');

1;
$ cat exporters/openssl.pc 
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: OpenSSL
Description: Secure Sockets Layer and cryptography libraries and tools
Version: 3.3.2-dev
Requires: libssl libcrypto
$ cat exporters/libssl.pc 
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: OpenSSL-libssl
Description: Secure Sockets Layer and cryptography libraries
Version: 3.3.2-dev
Requires.private: libcrypto
Libs: -L${libdir} -lssl
Cflags: -I${includedir}
$ cat exporters/libcrypto.pc 
prefix=/usr
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include
enginesdir=${libdir}/engines-3
modulesdir=${libdir}/ossl-modules

Name: OpenSSL-libcrypto
Description: OpenSSL cryptography library
Version: 3.3.2-dev
Libs: -L${libdir} -lcrypto
Libs.private: -ldl -pthread
Cflags: -I${includedir}
```

That being said, this reminds me of #23569...  could it be the same issue?  @icing, you were involved there, is this familiar to you?

**Comment by levitte** [comment]
Considering that the example installdata.pm in the top comment looks good, I assume the same goes for a run with the added patches, and that makes me suspect the templates in `exporters/pkg-config`.  There's a bit of perl code there, and it's *possible* that some perl versions give incorrect results (we've had run-ins with certain perl versions).

Which perl version did you use?

**Comment by Ra2-IFV** [comment]
```
$ perl -v

This is perl 5, version 36, subversion 0 (v5.36.0) built for x86_64-linux-gnu-thread-multi
(with 53 registered patches, see perl -V for more detail)

Copyright 1987-2022, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at https://www.perl.org/, the Perl Home Page.
```

**Comment by Ra2-IFV** [comment]
Using Debian 12 on x86 platform

**Comment by levitte** [comment]
And just in case, would you mind sharing the conf file that defines `linux-aarch64-openwrt`?  That one's not part of the regular OpenSSL dist

**Comment by Ra2-IFV** [comment]
I'm sorry but where is it? I don't know much about it...

**Comment by levitte** [comment]
`grep -n linux-aarch64-openwrt /home/debian/openwrt/build_dir/target-aarch64_generic_musl/openssl-3.3.1/Configurations`

That should tell you the file it's in.

Also, if you could share installdata.pm from the run with the #24687 patch?

**Comment by Ra2-IFV** [comment]
Okay, here is [installdata.pm.txt](https://github.com/user-attachments/files/15971335/installdata.pm.txt). Thanks for helping.
```
$ grep -nr linux-aarch64-openwrt /home/debian
/openwrt/build_dir/target-aarch64_generic_musl/openssl-3.3.1/Configurations/
/home/debian/openwrt/build_dir/target-aarch64_generic_musl/openssl-3.3.1/Configurations/25-openwrt.conf:10:    "linux-aarch64-openwrt" => {
```
```
## Openwrt "CONFIG_ARCH" matching targets.

# The targets need to end in '-openwrt' for the AFALG patch to work

my %targets = (
    "openwrt" => {
        template        => 1,
        CFLAGS          => add("\$(OPENWRT_OPTIMIZATION_FLAGS)"),
    },
    "linux-aarch64-openwrt" => {
        inherit_from    => [ "linux-aarch64", "openwrt" ],
    },
...OTHER PROFILE...
### Basic default option
    "linux-generic32-openwrt" => {
        inherit_from    => [ "linux-generic32", "openwrt" ],
    },
);
```
Generated by https://github.com/openwrt/openwrt/blob/main/package/libs/openssl/patches/110-openwrt_targets.patch

**Comment by levitte** [comment]
> Okay, here is [installdata.pm.txt](https://github.com/user-attachments/files/15971335/installdata.pm.txt).

Nope, that's not the right one.  If you look at my output, you'll see that the `_REL` variables have been renamed to `_REL_PREFIX` or `_REL_LIBDIR`.  If you have applied the patches from #24687, you should see those too.

**Comment by Ra2-IFV** [comment]
> Have to refresh [26d8638](https://github.com/openssl/openssl/commit/26d86382ea9ed144cae407437a2ac266f9146eee) to patch files successfully. Seems it's still broken...

I cloned your branch, with `--depth=3`, then I use `git format-patch -3 fix-24298` to create patches, and that's the problem... The oldest commit includes all the commits. How dumb I am... 
I fetched it one commit deeper and finally got the correct patch, it works as expected. Thank you very much for creating this pull request.

**Comment by levitte** [comment]
Hooray!
